shlock.1
.\" $Revision: 1.4.1 $
.TH SHLOCK 1
.SH NAME
shlock \- create lock files for use in shell scripts
.SH SYNOPSIS
.B shlock
[
.B \-h
]
[
.B \-V
]
[
.BI \-p " pid"
]
[
.B \-b
]
[
.B \-u
]
[
.B \-c
]
.BI \-f " name"
.SH DESCRIPTION
.I Shlock
tries to create a lock file named
.I name
and write the process ID
.I pid
into it.
If the file already exists,
.I shlock
will read the process ID from the file and test to see if the process
is currently active.
If the process exists, then the file will not be created.
.PP
Without
.B \-c
.I shlock
exits 0 (TRUE shell exit) if we obtained the lock on behalf of
.IR pid .
It exits 1 (FALSE shell exit) if already locked.
.PP
With
.B \-c
.I shlock
exits 0 (TRUE shell exit) if not locked, and
exits 1 (FALSE shell exit) if locked.
.PP
.B NOTE:
When we say the the is locked, we mean that
the file exists and contains the pid  of a currently-active process.
We do
.B NOT
refer to kernel file locking.
.PP
.B NOTE:
When a lock is obtained, it may be obtained
on behalf on when a lock is obtained, it may be obtained
on behalf of any process ID.
.PP
.B NOTE:
If a pid begins with
.I 0x
then it is assumed to be hex (base 16).
If a pid begins with
.I 0
then it is assumed to be in octal (base 8).
In all other cases the pid is assumed to be
in decimal (base 10).
The pid numeric value is assumed to
stop at the first non-digit, leading
.I 0x
not withstanding.
.SH OPTIONS
.TP
.B \-h
Print help message and exit.
.TP
.B \-V
Print version and exit.
.TP
.BI \-p " pid"
The process ID to write to the lock file if it is obtained.
By default,
.I pid
is the parent process ID of the
.I shlock
program.
That is, by default
.I pid
is the process ID of the program that forked
this tool.
.TP
.B \-b
Process IDs are normally read and written in ASCII.
If the ``\-b'' flag is used, then they will be written as a binary
.IR int .
.TP
.B \-u
For compatibility with other systems, the ``\-u'' flag is accepted as
a synonym for ``\-b'' since binary locks are used by many UUCP packages.
.TP
.B \-c
If the ``\-c'' flag is used, then
.I shlock
will not create a lock file, but will instead use the file to see if
the lock is held by another program.
.sp 1
If the lock is valid, the program will exit with a non-zero status; if
the lock is not valid (i.e., invoking
.I shlock
without the flag would have succeeded), then the program will exit
with a zero status.
When
.B \-c
is given, then any
.BI \-p " pid"
is ignored.
.TP
.BI \-f " name"
The lock file to create, test and/or use.
While this is an option, it is required.  :-)
.SH EXAMPLES
The following example shows how
.I shlock
would be used within a shell script:
.sp 1
.in +0.5i
.nf
# setup
#
SHLOCK=/usr/local/bin/shlock

# obtain the lock
#
# XXX - change the string prog to the basename of the script
LOCK=/tmp/.lock.prog
if [ ! -x "$SHLOCK" ]; then
    echo "$0: cannot find executable: $SHLOCK" 1>&2
    exit 1
fi
if ! "$SHLOCK" -p $$ -f "$LOCK"; then
    echo "$0: locked by process" `cat "$LOCK"` 1>&2
    exit 2
fi
# must trap after locking to avoid removal of another process lock
trap "rm -f $LOCK; exit" 0 1 2 3 15

# do something ...
.fi
.in -0.5i
.sp 1
In another example, let us assume we want to write a script with
an
.I \-F
option that indicated that the script should proceed even if
the lock was not granted.
That is
.I \-F
means to force past a lock.
Here is how this would be done:
.sp 1
.in +0.5i
.nf
# setup
#
SHLOCK=/usr/local/bin/shlock

USAGE="usage: $0 [-F] arg ..."
F_FLAG=
set -- `/usr/bin/getopt F $*`
if [ $? != 0 ]; then
    echo "$0: unknown or invalid -flag" 1>&2
    echo $USAGE 1>&2
    exit 1
fi
for i in $*; do
    case "$i" in
    -F) F_FLAG="true" ;;
    esac
    shift
done
if [ $# -lt 1 ]; then
    echo "$0: must have at least one arg" 1>&2
    echo $USAGE 1>&2
    exit 2
fi

# ...

# obtain the lock
#
# XXX - change the string prog to the basename of the script
LOCK=/tmp/.lock.progname
if [ ! -x "$SHLOCK" ]; then
    echo "$0: cannot find executable: $SHLOCK" 1>&2
    exit 3
fi
if ! "$SHLOCK" -p $$ -f "$LOCK"; then
    if [ -z "$FORCE" ]; then
	echo "$0: locked by process" `cat "$LOCK"` 1>&2
	exit 2
    else
    	# the -F forced a lock override, so we must not remove the lock
	LOCK=
    fi
fi
# must trap after locking to avoid removal of another process lock
trap "rm -f $LOCK; exit" 0 1 2 3 15

# do something ...
.fi
.in -0.5i
.sp 1
In this example, we want silently exit if there is lock
and proceed otherwise.
We assume that something else will sometimes hold the lock.
We do not want to lock it ourselves, only ensure that
the lock is not held by someone else.
Here is how this would be done:
.sp 1
.in +0.5i
.nf
# setup
#
SHLOCK=/usr/local/bin/shlock

# ensure that it is not locked
#
LOCK=/tmp/.lock.name
if [ ! -x "$SHLOCK" ]; then
    echo "$0: cannot find executable: $SHLOCK" 1>&2
    exit 1
fi
if ! "$SHLOCK" -c -f "$LOCK"; then
    # locked is else by someone, so silently exit
    exit 0
fi

# do something ...

.fi
.in -0.5i
.SH HISTORY
Written by Rich $alz <rsalz@uunet.uu.net> after a description of HDB UUCP
locking given by Peter Honeyman.
.sp
Bug fixes relating to
.I \-c
and the removal a needless sleep by
Landon Curt Noll
(http://www.isthe.com/chongo/index.html).
.de R$
.sp 1
This is revision \\$3, dated \\$4.
..
.R$ $Id: shlock.1,v 1.4.1 2025/03/25 09:28:02 chongo Exp $


shlock.c
/*
**  Produce reliable locks for shell scripts, by Peter Honeyman as told
**  to Rich $alz.
**
** @(#) $Revision: 1.6 $
** @(#) $Id: shlock.c,v 1.6 2005/01/05 09:28:02 chongo Exp $
** @(#) $Source: /usr/local/src/bin/shlock/RCS/shlock.c,v $
*/

#include <stdio.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/types.h>
#include <signal.h>
#include <sys/stat.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdbool.h>

/*
 * official version
 */
#define VERSION "1.6.1 2025-03-25"          /* format: major.minor YYYY-MM-DD */

#define NEWSUMASK 0002
#define MAX_TMP_TRY 17		/* max tmp file open retries */
#define TMP_USEC_WAIT 1270000	/* micro secs beteen tmp open or link retries */

static bool	BinaryLock;
static char	CANTUNLINK[] = "Can't unlink \"%s\", %s\n";
static char	CANTOPEN[] = "Can't open \"%s\", %s\n";


/*
**  See if the process named in an existing lock still exists by
**  sending it a null signal.
*/
static bool
ValidLock(char *name, bool JustChecking)
{
    register int	fd;
    register int	i;
    long		pid;
    char		buff[BUFSIZ];

    /* Open the file. */
    errno = 0;
    if ((fd = open(name, O_RDONLY)) < 0) {
	if (JustChecking) {
	    return false;
	} else {
	    (void)fprintf(stderr, CANTOPEN, name, strerror(errno));
	    return true;
	}
    }

    /* Read the PID that is written there. */
    if (BinaryLock) {
	if (read(fd, (char *)&pid, sizeof pid) != sizeof pid) {
	    (void)close(fd);
	    return false;
	}
    }
    else {
	if ((i = read(fd, buff, sizeof buff - 1)) <= 0) {
	    (void)close(fd);
	    return false;
	}
	buff[i] = '\0';
	errno = 0;
	pid = strtol(buff, NULL, 0);
	if (errno != 0) {
	    (void)close(fd);
	    return false;
	}
    }
    (void)close(fd);
    if ((pid_t)pid <= 0)
	return false;

    /* Send the signal. */
    errno = 0;
    if (kill((pid_t)pid, 0) < 0 && errno == ESRCH)
	return false;

    /* Either the kill worked, or we're optimistic about the error code. */
    return true;
}


/*
**  Unlink a file, print a message on error, and exit.
*/
static void
UnlinkAndExit(char *name, int x)
{
    errno = 0;
    if (unlink(name) < 0)
	(void)fprintf(stderr, CANTUNLINK, name, strerror(errno));
    exit(x);
}


/*
**  Print a usage message and exit.
*/
static void
Usage(void)
{
    (void)fprintf(stderr,
        "Usage: shlock [-h] [-V] [-p pid] [-b|-u] [-c] -f file\n"
	"\n"
	"\t-h\t\tprint help message and exit\n"
	"\t-V\t\tprint version string and exit\n"
	"\n"
	"\t-p pid\t\tif lock is granted, lock with pid (def: parent process ID)\n"
	"\t-b\t\twrite pid into lock file in binary format (def: ASCII)\n"
	"\t-u\t\talias for -b\n"
	"\t-c\t\ttest for lock without locking\n"
	"\n"
	"\t-f file\t\tlock filename (NOTE: -f file is required)\n"
	"\n"
	"without -c:\n"
	"\n"
	"\texit 0 (true shell exit) if lock was created and granted to pid\n"
	"\texit 1 (false shell exit) if locked\n"
	"\n"
	"with -c:\n"
	"\n"
	"\texit 0 (true shell exit) if not locked\n"
	"\texit 1 (false shell exit) if locked\n"
	"\n"
	"shlock version: %s\n", VERSION);
    exit(1);
    /* NOTREACHED */
}


int
main(int ac, char *av[])
{
    register int	i;
    register char	*p;
    register int	fd;
    char		tmp[BUFSIZ+1];
    char		buff[BUFSIZ+1];
    char		*name;
    long		pid;
    bool		ok;
    bool		JustChecking;
    extern int		optind;
    extern char *	optarg;
    char *program;

    /* Set defaults. */
    program = av[0];
    pid = 0;
    name = NULL;
    JustChecking = false;
    (void)umask(NEWSUMASK);
    pid = (long)getppid();
    memset(tmp, 0, sizeof(tmp));
    memset(buff, 0, sizeof(buff));

    /* Parse JCL. */
    while ((i = getopt(ac, av, "hVbcup:f:")) != EOF)
	switch (i) {
	case 'h':                   /* -h - print help message and exit */
	    Usage();
	    /* NOTREACHED */

	case 'V':
	    (void) printf("%s\n", VERSION);
	    exit(2); /* ooo */
	    /*NOTREACHED*/

	case 'b':
	case 'u':
	    BinaryLock = true;
	    break;

	case 'c':
	    JustChecking = true;
	    break;

	case 'p':
	    errno = 0;
	    pid = strtol(optarg, NULL, 0);
	    if (errno != 0) {
		pid = 0;
	    }
	    break;

	case 'f':
	    name = optarg;
	    break;

	case ':':
            (void) fprintf(stderr, "%s: ERROR: requires an argument -- %c\n", program, optopt);
	    Usage();
            /*NOTREACHED*/

        case '?':
            (void) fprintf(stderr, "%s: ERROR: illegal option -- %c\n", program, optopt);
	    Usage();
            /*NOTREACHED*/

	default:
	    Usage();
	    /* NOTREACHED */
	    break;
	}
    ac -= optind;
    av += optind;
    if (ac || (pid_t)pid <= 0 || name == NULL)
	Usage();

    /* Create the temp file in the same directory as the destination. */
    if ((p = strrchr(name, '/')) != NULL) {
	*p = '\0';
	(void)sprintf(tmp, "%s/shlock%ld", name, (long)getpid());
	*p = '/';
    }
    else
	(void)sprintf(tmp, "shlock%ld", (long)getpid());

    /* Loop for a while, until we can open the file. */
    i = 0;
    errno = 0;
    while ((fd = open(tmp, O_RDWR | O_CREAT | O_EXCL, 0644)) < 0) {

	/* process open failure */
	switch (errno) {
	default:
	    /* Unknown error -- give up. */
	    (void)fprintf(stderr, CANTOPEN, tmp, strerror(errno));
	    exit(1);
	case EEXIST:
	    /* If we can remove the old temporary, retry the open. */
	    if (unlink(tmp) < 0) {
		(void)fprintf(stderr, CANTUNLINK, tmp, strerror(errno));
		exit(1);
	    }
	    break;
	}

	/* wait and retry */
	if (++i >= MAX_TMP_TRY) {
	    (void)fprintf(stderr, CANTOPEN, tmp, "too many retries");
	    exit(1);
	}

	/* wait before reattempting to get tmp file */
	usleep(TMP_USEC_WAIT);
	errno = 0;
    }

    /* Write the process ID. */
    errno = 0;
    if (BinaryLock)
	ok = write(fd, (void *)pid, (size_t)sizeof pid) == sizeof pid;
    else {
	(void)sprintf(buff, "%ld\n", pid);
	i = strlen(buff);
	ok = write(fd, (void *)buff, (size_t)i) == i;
    }
    if (!ok) {
	(void)fprintf(stderr, "Can't write PID to \"%s\", %s\n",
	    tmp, strerror(errno));
	(void)close(fd);
	UnlinkAndExit(tmp, 1);
    }

    (void)close(fd);

    /* Handle the "-c" flag. */
    if (JustChecking) {
	if (ValidLock(name, true))
	    UnlinkAndExit(tmp, 1);
	UnlinkAndExit(tmp, 0);
    }

    /* Try to link the temporary to the lockfile. */
    i = 0;
    errno = 0;
    while (link(tmp, name) < 0) {
	switch (errno) {
	case EEXIST:
	    /* File exists; if lock is valid, give up. */
	    if (ValidLock(name, false))
		UnlinkAndExit(tmp, 1);
	    if (unlink(name) < 0) {
		(void)fprintf(stderr, CANTUNLINK, name, strerror(errno));
		UnlinkAndExit(tmp, 1);
	    }
	    break;
	default:
	    /* Unknown error -- give up. */
	    (void)fprintf(stderr, "Can't link \"%s\" to \"%s\", %s\n",
		    tmp, name, strerror(errno));
	    UnlinkAndExit(tmp, 1);
	    break;
	}

	/* wait and retry */
	if (++i >= MAX_TMP_TRY) {
	    (void)fprintf(stderr, "Too many link retries of \"%s\" to \"%s\"\n",
			  tmp, name);
	    UnlinkAndExit(tmp, 1);
	    /* NOTREACHED */
	}

	/* wait before reattempting to get relink the tmp file */
	if (i > 1) {
	    usleep(TMP_USEC_WAIT);
	}
	errno = 0;
    }

    UnlinkAndExit(tmp, 0);
    /* NOTREACHED */
    exit(0);
}
